Source/Scene/ShadowMap.js

```diff
@@ -390,67 +411,67 @@ define([
             destroyFramebuffer(shadowMap);
             createFramebuffer(shadowMap, context);
         }
     }
 
-    ShadowMap.prototype.setNumberOfCascades = function(numberOfCascades) {
-        //>>includeStart('debug', pragmas.debug);
-        if (!defined(numberOfCascades) || ((numberOfCascades !== 1) && (numberOfCascades !== 4))) {
-            throw new DeveloperError('Only one or four cascades are supported.');
-        }
-        //>>includeEnd('debug');
-
-        this._numberOfCascades = numberOfCascades;
-
-        if (numberOfCascades === 1) {
-            this._cascadeViewports[0] = new Cartesian4(0.0, 0.0, 1.0, 1.0);
-        } else if (numberOfCascades === 4) {
-            // Mirrors the viewports in the shader (offset x, offset y, scale x, scale y)
-            this._cascadeViewports[0] = new Cartesian4(0.0, 0.0, 0.5, 0.5);
-            this._cascadeViewports[1] = new Cartesian4(0.5, 0.0, 0.5, 0.5);
-            this._cascadeViewports[2] = new Cartesian4(0.0, 0.5, 0.5, 0.5);
-            this._cascadeViewports[3] = new Cartesian4(0.5, 0.5, 0.5, 0.5);
-        }
-    };
-
     ShadowMap.prototype.setSize = function(size) {
-        if (this._numberOfCascades === 4) {
-            size *= 2;
-        }
-
         this._shadowMapSize = size;
-
-        // Update pass state
-        this._passState.viewport.x = 0;
-        this._passState.viewport.y = 0;
-        this._passState.viewport.width = size;
-        this._passState.viewport.height = size;
-
-        // Update cascade pass states
-        for (var i = 0; i < this._numberOfCascades; ++i) {
-            var passState = this._cascadePassStates[i];
-            var viewport = this._cascadeViewports[i];
-            passState.viewport.x = size * viewport.x;
-            passState.viewport.y = size * viewport.y;
-            passState.viewport.width = size * viewport.z;
-            passState.viewport.height = size * viewport.w;
+        var numberOfPasses = this._numberOfPasses;
+        var textureSize = this._textureSize;
+
+        if (numberOfPasses === 1) {
+            // +----+
+            // |  1 |
+            // +----+
+            textureSize.x = size;
+            textureSize.y = size;
+            this._passStates[0].viewport = new BoundingRectangle(0, 0, size, size);
+        } else if (numberOfPasses === 4) {
+            // +----+----+
+            // |  3 |  4 |
+            // +----+----+
+            // |  1 |  2 |
+            // +----+----+
+            textureSize.x = size * 2;
+            textureSize.y = size * 2;
+            this._passStates[0].viewport = new BoundingRectangle(0, 0, size, size);
+            this._passStates[1].viewport = new BoundingRectangle(size, 0, size, size);
+            this._passStates[2].viewport = new BoundingRectangle(0, size, size, size);
+            this._passStates[3].viewport = new BoundingRectangle(size, size, size, size);
+        } else if (numberOfPasses === 6) {
+            // +----+----+----+
```

Comment:
Also, as we discussed offline, we are going to have cases like spot lights where we know we won't need all 6 faces.  At the least, we should be able to optimize the number of passes, but if we think it is worth it, I suppose we can also optimize the texture size (but not for dynamic field of views).